diff --git m4/check-libc.m4 m4/check-libc.m4
index e511f6d..d11493f 100644
--- m4/check-libc.m4
+++ m4/check-libc.m4
@@ -115,7 +115,7 @@ AC_CACHE_CHECK([for getentropy], ac_cv_func_getentropy, [
 	])
 ])
 
-AC_CHECK_FUNCS([timingsafe_bcmp timingsafe_memcmp])
+AC_CHECK_FUNCS([timingsafe_memcmp])
 AM_CONDITIONAL([HAVE_ARC4RANDOM], [test "x$ac_cv_func_arc4random" = xyes])
 AM_CONDITIONAL([HAVE_ARC4RANDOM_BUF], [test "x$ac_cv_func_arc4random_buf" = xyes])
 AM_CONDITIONAL([HAVE_ARC4RANDOM_UNIFORM], [test "x$ac_cv_func_arc4random_uniform" = xyes])
